src/python/CMSSpark/data_aggregation.py

```diff
@@ -0,0 +1,385 @@
+
+
+    # Build a query with "cols" columns
+    query = ("SELECT %s FROM aaa_df " \
+             "JOIN f_b_s_df ON f_b_s_df.file_name = aaa_df.file_lfn " \
+             "GROUP BY src_experiment_site, dataset_name") % ','.join(cols)
+
+    result = run_query(query, sql_context, verbose)
+
+    result = result.sort(desc("nacc"))
+
+    if verbose:
+        print 'Finished AAA part (output is ' + str(result.count()) + ' records)'
+
+    return result
+
+
+def run_agg_cmssw(date, fout, ctx, sql_context, verbose=False):
+
+    if verbose:
+        print 'Starting CMSSW part'
+
+    # Convert date
+    date = long_date_string(date)
+
+    # Create CMSSW tables in sql_context
+    cmssw_df = cmssw_tables(ctx, sql_context, date=date, verbose=verbose)
+
+    if verbose:
+        print 'Found ' + str(cmssw_df['cmssw_df'].count()) + ' records in CMSSW stream'
+
+    # - site name                +
+    # - dataset name             +
+    # - number of access (nacc)  +
+    # - distinct users           +
+    # - stream: cmssw            +
+
+    cols = ['cmssw_df.SITE_NAME AS site_name',
+            'dataset_name',
+            'count(dataset_name) AS nacc',
+            'count(distinct(USER_DN)) AS distinct_users',
+            '\"cmssw\" as stream']
+
+    # Build a query with "cols" columns
+    query = ("SELECT %s FROM cmssw_df "\
+             "JOIN f_b_s_df ON f_b_s_df.file_name = cmssw_df.FILE_LFN "\
+             "GROUP BY cmssw_df.SITE_NAME, dataset_name") % ','.join(cols)
+
+    result = run_query(query, sql_context, verbose)
+
+    result = result.sort(desc("nacc"))
+
+    if verbose:
+        print 'Finished CMSSW part (output is ' + str(result.count()) + ' records)'
+
+    return result
+
+
+def quiet_logs( sc ):
+    print '*** WILL QUIET LOGS ***'
+    logger = sc._jvm.org.apache.log4j
+    logger.LogManager.getRootLogger().setLevel(logger.Level.ERROR)
+    print '*** DID QUIET LOGS ***'
+
+
+def create_file_block_site_table(ctx, sql_context, verbose=False):
+    if verbose:
+        print 'Starting file_block_site generation'
+
+    cols = ['f_logical_file_name AS file_name',
+            'b_block_name AS block_name',
+            'clean_site_name(node_name) AS site_name',
+            'd_dataset AS dataset_name']
+
+    # Join FDF and BDF by f_block_id and b_block_id
+    query = ("SELECT %s FROM fdf " \
+             "JOIN bdf ON fdf.f_block_id = bdf.b_block_id "\
+             "JOIN ddf ON fdf.f_dataset_id = ddf.d_dataset_id "\
+             "JOIN phedex_df ON bdf.b_block_name = phedex_df.block_name") % ','.join(cols)
+
+    if verbose:
+        print 'Will run query to generate temp file_block_site table'
+
+    result = run_query(query, sql_context, verbose)
+    result.registerTempTable('f_b_all_df')
+
+    query_distinct = ("SELECT DISTINCT * FROM f_b_all_df ORDER  BY file_name")
+    result_distinct = run_query(query_distinct, sql_context, verbose)
+    result_distinct.registerTempTable('f_b_s_df')
+
```

Comment:
In my `run_query` function (which is in `data_collection.py`) I do `persist(...)` for all dataframes that are created there.